src/ripple/basics/PerfLog.h

```diff
@@ -0,0 +1,171 @@
+    using milliseconds = std::chrono::milliseconds;
+    using microseconds = std::chrono::microseconds;
+
+    /**
+     * Configuration from [perf] section of rippled.cfg.
+     */
+    struct Setup
+    {
+        boost::filesystem::path perf_log;
+        // log_interval is in milliseconds to support faster testing.
+        milliseconds log_interval {seconds(1)};
+    };
+
+    virtual ~PerfLog()
+    { }
+
+    /**
+     * Log start of RPC call.
+     *
+     * @param method RPC command
+     * @param requestId Unique identifier to track command
+     */
+    virtual void rpcStart(std::string const& method,
+        std::uint64_t requestId) = 0;
+
+    /**
+     * Log successful finish of RPC call
+     *
+     * @param method RPC command
+     * @param requestId Unique identifier to track command
+     */
+    virtual void rpcFinish(std::string const& method,
+        std::uint64_t requestId) = 0;
+
+    /**
+     * Log errored RPC call
+     *
+     * @param method RPC command
+     * @param requestId Unique identifier to track command
+     */
+    virtual void rpcError(std::string const& method,
+        std::uint64_t requestId) = 0;
+
+    /**
+     * Log queued job
+     *
+     * @param type Job type
+     */
+    virtual void jobQueue(JobType const type) = 0;
+
+    /**
+     * Log job executing
+     *
+     * @param type Job type
+     * @param dur Duration enqueued in microseconds
+     * @param startTime Time that execution began
+     * @param instance JobQueue worker thread instance
+     */
+    virtual void jobStart(JobType const type,
+        microseconds dur,
+        steady_time_point startTime,
+        int instance) = 0;
+
+    /**
+     * Log job finishing
+     *
+     * @param type Job type
+     * @param dur Duration running in microseconds
+     * @param instance Jobqueue worker thread instance
+     */
+    virtual void jobFinish(JobType const type,
+        microseconds dur, int instance) = 0;
+
+    /**
+     * Render performance counters in Json
+     *
+     * @return Counters Json object
+     */
+    virtual Json::Value countersJson() const = 0;
+
+    /**
+     * Render currently executing jobs and RPC calls and durations in Json
+     *
+     * @return Current executing jobs and RPC calls and durations
+     */
+    virtual Json::Value currentJson() const = 0;
+
+    /**
+     * Ensure enough room to store each currently executing job
+     *
+     * @param resize Number of JobQueue worker threads
+     */
+    virtual void resizeJobs(int const resize) = 0;
+
+    /**
+     * Rotate perf log file
+     */
+    virtual void rotate() = 0;
+};
+
+} //perf
+
+class Section;
+class Stoppable;
+
+namespace perf {
+
+PerfLog::Setup setup_PerfLog(Section const& section,
+    boost::filesystem::path const& CONFIG_DIR);
+
+std::unique_ptr<PerfLog> make_PerfLog(
+    PerfLog::Setup const& setup,
+    Stoppable& parent,
+    beast::Journal journal,
+    std::function<void()>&& signalStop);
```

Comment:
That would work too.  I coded it as an rvalue ref because that's what is expected at this point.  There are two call sites: one in Application and one in the unit tests.  Both of those call sites pass lambdas that get converted into an rvalue `std::function`.  There is a small bit of maintenance if a value is passed, since you must remember to move from it or else you get a copy.  But I'm happy with either approach.